src/states/establishing_node.rs

```diff
@@ -0,0 +1,463 @@
+    fmt::{self, Display, Formatter},
+};
+
+const POKE_TIMEOUT: Duration = Duration::from_secs(60);
+
+pub struct EstablishingNode {
+    ack_mgr: AckManager,
+    cache: Box<Cache>,
+    chain: Chain,
+    crust_service: Service,
+    full_id: FullId,
+    gen_pfx_info: GenesisPfxInfo,
+    /// Routing messages addressed to us that we cannot handle until we are established.
+    msg_backlog: Vec<RoutingMessage>,
+    notified_nodes: BTreeSet<PublicId>,
+    parsec: Parsec,
+    peer_mgr: PeerManager,
+    poke_timer_token: u64,
+    routing_msg_filter: RoutingMessageFilter,
+    timer: Timer,
+}
+
+impl EstablishingNode {
+    #[allow(clippy::too_many_arguments)]
+    pub fn from_proving_node(
+        ack_mgr: AckManager,
+        cache: Box<Cache>,
+        crust_service: Service,
+        full_id: FullId,
+        gen_pfx_info: GenesisPfxInfo,
+        min_section_size: usize,
+        msg_backlog: Vec<RoutingMessage>,
+        notified_nodes: BTreeSet<PublicId>,
+        peer_mgr: PeerManager,
+        routing_msg_filter: RoutingMessageFilter,
+        timer: Timer,
+    ) -> Self {
+        let public_id = *full_id.public_id();
+        let poke_timer_token = timer.schedule(POKE_TIMEOUT);
+
+        let parsec = parsec::create(full_id.clone(), &gen_pfx_info);
+        let chain = Chain::new(min_section_size, public_id, gen_pfx_info.clone());
+
+        let node = Self {
+            ack_mgr,
+            cache,
+            chain,
+            crust_service,
+            full_id,
+            gen_pfx_info,
+            msg_backlog,
+            notified_nodes,
+            parsec,
+            peer_mgr,
+            poke_timer_token,
+            routing_msg_filter,
+            timer,
+        };
+
+        debug!("{} - State changed to EstablishingNode.", node);
+        node
+    }
+
+    pub fn into_node(
+        self,
+        sec_info: SectionInfo,
+        old_pfx: Prefix<XorName>,
+        outbox: &mut EventBox,
+    ) -> State {
+        let node = Node::from_establishing_node(
+            self.ack_mgr,
+            self.cache,
+            self.chain,
+            self.crust_service,
+            self.full_id,
+            self.gen_pfx_info,
+            self.msg_backlog.into_iter().collect(),
+            self.notified_nodes,
+            old_pfx,
+            self.parsec,
+            self.peer_mgr,
+            self.routing_msg_filter,
+            sec_info,
+            self.timer,
+            outbox,
+        );
+
+        match node {
+            Ok(node) => State::Node(node),
+            Err(_) => State::Terminated,
+        }
+    }
+
+    #[cfg(feature = "mock")]
+    pub fn chain(&self) -> &Chain {
+        &self.chain
+    }
+
+    fn dispatch_routing_message(
+        &mut self,
+        msg: RoutingMessage,
+        outbox: &mut EventBox,
+    ) -> Result<Transition, RoutingError> {
+        self.handle_routing_message(msg, outbox)
+            .map(|()| Transition::Stay)
+    }
+
+    fn handle_parsec_request(
```

Comment:
Maybe as a follow up commit remove duplication with version in node, and code in parsec.rs.

Ideally the following would not be duplicated.

```rust
parsec.handle_request(&pub_id, request)
            .map(|response| response.into_message(msg_version))
            .map_err(|err| {
                debug!("{} - Error handling parsec request: {:?}", log_ident, err);
                err
            })
            .ok();
```

Same for parsec.handle_response.

Ideally the code calling handle_request followed by parsec_poll_all would also not be duplicated.

Reply:
I wonder if we would not be better off keeping the member as a ParsecMap, then this issue would not exists, and have a `take_sole_parsec` for the into_...

Reply:
as it happen, with NodeAgeing, we will probably stay in establishing through multiple Parsec, so having the map may be good?

Reply:
(If we go back to a map, better to do it in that initial commit,

Reply:
The reason I was suggesting this, is primarily because that it is a much simpler diff if we use ParsecMap everywhere (since then there is no need for duplicate code or de-duplication, the code is there). The added benefit is it also seem to be closer to what we will want.

Reply:
OK, I think you might be right. I'll give it a try.